Regression in 2.1.214 when joining 2 recursive CTE containing bind values

I have an integration test that uses 2 recursive CTE and joins them like this:

```sql
with recursive
  t1(f1, f2) as (
    select
      1,
      cast('a' as varchar(15))
    union all
    select
      (t1.f1 + 1),
      cast(('a' || t1.f2) as varchar(15))
    from t1
    where t1.f1 < 10
  ),
  t2(g1, g2) as (
    select
      1,
      cast('b' as varchar(15))
    union all
    select
      (t2.g1 + 1),
      cast(('b' || t2.g2) as varchar(15))
    from t2
    where t2.g1 < 10
  )
select t1.f1, t1.f2, t2.g1, t2.g2
from t1
  join t2
    on t1.f1 = t2.g1
``` 

The result being:

```
|F1 |F2        |G1 |G2        |
|---|----------|---|----------|
|1  |a         |1  |b         |
|2  |aa        |2  |bb        |
|3  |aaa       |3  |bbb       |
|4  |aaaa      |4  |bbbb      |
|5  |aaaaa     |5  |bbbbb     |
|6  |aaaaaa    |6  |bbbbbb    |
|7  |aaaaaaa   |7  |bbbbbbb   |
|8  |aaaaaaaa  |8  |bbbbbbbb  |
|9  |aaaaaaaaa |9  |bbbbbbbbb |
|10 |aaaaaaaaaa|10 |bbbbbbbbbb|
```

However, when I use bind variables, it seems like something goes wrong:

```java
try (PreparedStatement s = connection.prepareStatement("""
    with recursive
      t1(f1, f2) as (
        select
          cast(? as int),
          cast(? as varchar(15))
        union all
        select
          (t1.f1 + cast(? as int)),
          cast((cast(? as varchar) || t1.f2) as varchar(15))
        from t1
        where t1.f1 < 10
      ),
      t2(g1, g2) as (
        select
          cast(? as int),
          cast(? as varchar(15))
        union all
        select
          (t2.g1 + 1),
          cast(('b' || t2.g2) as varchar(15))
        from t2
        where t2.g1 < 10
      )
    select t1.f1, t1.f2, t2.g1, t2.g2
    from t1
      join t2
        on t1.f1 = t2.g1
    """)) {
    int i = 1;
    s.setInt(i++, 1);
    s.setString(i++, "a");
    s.setInt(i++, 1);
    s.setString(i++, "a");
    s.setInt(i++, 1);
    s.setString(i++, "b");

    try (ResultSet rs = s.executeQuery()) {
        while (rs.next()) {
            System.out.println(rs.getString(2));
        }
    }
}
```

I'm getting this error now with H2 2.1.214:

```
General error: "java.lang.IndexOutOfBoundsException: Index 4 out of bounds for length 2"; SQL statement:
with recursive
  t1(f1, f2) as (
    select
      cast(? as int),
      cast(? as varchar(15))
    union all
    select
      (t1.f1 + cast(? as int)),
      cast((cast(? as varchar) || t1.f2) as varchar(15))
    from t1
    where t1.f1 < 10
  ),
  t2(g1, g2) as (
    select
      cast(? as int),
      cast(? as varchar(15))
    union all
    select
      (t2.g1 + 1),
      cast(('b' || t2.g2) as varchar(15))
    from t2
    where t2.g1 < 10
  )
select t1.f1, t1.f2, t2.g1, t2.g2
from t1
  join t2
    on t1.f1 = t2.g1 [50000-214]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:554)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)
	at org.h2.message.DbException.get(DbException.java:212)
	at org.h2.message.DbException.convert(DbException.java:395)
	at org.h2.command.Command.executeQuery(Command.java:211)
	at org.h2.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:128)
	at org.jooq.testscripts.JDBC.jdbc(JDBC.java:98)
	at org.jooq.testscripts.JDBC.main(JDBC.java:43)
Caused by: java.lang.IndexOutOfBoundsException: Index 4 out of bounds for length 2
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:359)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at org.h2.command.Parser.readParameter(Parser.java:4877)
	at org.h2.command.Parser.readTermWithoutIdentifier(Parser.java:4950)
	at org.h2.command.Parser.readTerm(Parser.java:4901)
	at org.h2.command.Parser.readFactor(Parser.java:3398)
	at org.h2.command.Parser.readSum(Parser.java:3385)
	at org.h2.command.Parser.readConcat(Parser.java:3350)
	at org.h2.command.Parser.readCondition(Parser.java:3132)
	at org.h2.command.Parser.readExpression(Parser.java:3053)
	at org.h2.command.Parser.readTermWithoutIdentifier(Parser.java:5105)
	at org.h2.command.Parser.readTerm(Parser.java:4901)
	at org.h2.command.Parser.readFactor(Parser.java:3398)
	at org.h2.command.Parser.readSum(Parser.java:3385)
	at org.h2.command.Parser.readConcat(Parser.java:3350)
	at org.h2.command.Parser.readCondition(Parser.java:3132)
	at org.h2.command.Parser.readExpression(Parser.java:3053)
	at org.h2.command.Parser.parseSelectExpressions(Parser.java:2853)
	at org.h2.command.Parser.parseSelect(Parser.java:2871)
	at org.h2.command.Parser.parseQueryPrimary(Parser.java:2762)
	at org.h2.command.Parser.parseQueryTerm(Parser.java:2633)
	at org.h2.command.Parser.parseQueryExpressionBody(Parser.java:2612)
	at org.h2.command.Parser.parseQueryExpressionBodyAndEndOfQuery(Parser.java:2605)
	at org.h2.command.Parser.parseQueryPrimary(Parser.java:2756)
	at org.h2.command.Parser.parseQueryTerm(Parser.java:2633)
	at org.h2.command.Parser.parseQueryExpressionBody(Parser.java:2612)
	at org.h2.command.Parser.parseQueryExpressionBodyAndEndOfQuery(Parser.java:2605)
	at org.h2.command.Parser.parseQueryExpression(Parser.java:2598)
	at org.h2.command.Parser.parseQuery(Parser.java:2567)
	at org.h2.command.Parser.parsePrepared(Parser.java:724)
	at org.h2.command.Parser.parse(Parser.java:689)
	at org.h2.command.Parser.parse(Parser.java:661)
	at org.h2.command.Parser.prepare(Parser.java:537)
	at org.h2.index.QueryExpressionIndex.findRecursive(QueryExpressionIndex.java:171)
	at org.h2.index.QueryExpressionIndex.find(QueryExpressionIndex.java:266)
	at org.h2.index.QueryExpressionIndex.find(QueryExpressionIndex.java:152)
	at org.h2.index.IndexCursor.find(IndexCursor.java:161)
	at org.h2.table.TableFilter.next(TableFilter.java:394)
	at org.h2.table.TableFilter.next(TableFilter.java:464)
	at org.h2.command.query.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1832)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:78)
	at org.h2.result.FetchedResult.next(FetchedResult.java:34)
	at org.h2.command.query.Select.queryFlat(Select.java:728)
	at org.h2.command.query.Select.queryWithoutCache(Select.java:833)
	at org.h2.command.query.Query.queryWithoutCacheLazyCheck(Query.java:197)
	at org.h2.command.query.Query.query(Query.java:512)
	at org.h2.command.query.Query.query(Query.java:475)
	at org.h2.command.CommandContainer.query(CommandContainer.java:251)
	at org.h2.command.Command.executeQuery(Command.java:190)
	... 3 more
```

This seems to be a regression. With H2 2.1.212, the output was as expected:

```
a
aa
aaa
aaaa
aaaaa
aaaaaa
aaaaaaa
aaaaaaaa
aaaaaaaaa
aaaaaaaaaa
```

The workaround seems to be to avoid bind values in CTE.

This change might have caused it? https://github.com/h2database/h2database/pull/3539/commits/90caa1d36a21fc0b0b58e2fbefd83e46acff9bd0

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,10 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #3554: Regression in 2.1.214 when joining 2 recursive CTE containing bind values
+</li>
+<li>PR #3989: Refactor CTE-related code and reduce recompilations
+</li>
 <li>Issue #3987: Allow empty &lt;with column list&gt;
 </li>
 <li>Issue #822: WITH clauses' column aliases are not maintained correctly when selecting from CTE from within a derived

--- a/h2/src/main/org/h2/command/ParserBase.java
+++ b/h2/src/main/org/h2/command/ParserBase.java
@@ -217,7 +217,19 @@ public class ParserBase {
     }
 
     public final void setSuppliedParameters(ArrayList<Parameter> suppliedParameters) {
-        this.parameters = suppliedParameters;
+        int max = Parameter.getMaxIndex(suppliedParameters);
+        if (max > suppliedParameters.size()) {
+            ArrayList<Parameter> parameters = new ArrayList<>(max);
+            for (int i = 0; i < max; i++) {
+                parameters.add(null);
+            }
+            for (Parameter p : suppliedParameters) {
+                parameters.set(p.getIndex(), p);
+            }
+            this.parameters = parameters;
+        } else {
+            this.parameters = suppliedParameters;
+        }
     }
 
     /**

--- a/h2/src/main/org/h2/expression/Parameter.java
+++ b/h2/src/main/org/h2/expression/Parameter.java
@@ -5,6 +5,8 @@
  */
 package org.h2.expression;
 
+import java.util.ArrayList;
+
 import org.h2.api.ErrorCode;
 import org.h2.engine.SessionLocal;
 import org.h2.expression.condition.Comparison;
@@ -19,6 +21,26 @@ import org.h2.value.ValueVarchar;
  * A parameter of a prepared statement.
  */
 public final class Parameter extends Operation0 implements ParameterInterface {
+
+    /**
+     * Returns the maximum 1-based index.
+     *
+     * @param parameters
+     *            parameters
+     * @return the maximum 1-based index, or {@code -1}
+     */
+    public static int getMaxIndex(ArrayList<Parameter> parameters) {
+        int result = 0;
+        for (Parameter p : parameters) {
+            if (p != null) {
+                int index = p.getIndex() + 1;
+                if (index > result) {
+                    result = index;
+                }
+            }
+        }
+        return result;
+    }
 
     private Value value;
     private Column column;

--- a/h2/src/main/org/h2/table/QueryExpressionTable.java
+++ b/h2/src/main/org/h2/table/QueryExpressionTable.java
@@ -105,16 +105,6 @@ public abstract class QueryExpressionTable extends Table {
         return columnTemplateList;
     }
 
-    static int getMaxParameterIndex(ArrayList<Parameter> parameters) {
-        int result = -1;
-        for (Parameter p : parameters) {
-            if (p != null) {
-                result = Math.max(result, p.getIndex());
-            }
-        }
-        return result;
-    }
-
     Query viewQuery;
 
     ArrayList<Table> tables;
@@ -249,11 +239,11 @@ public abstract class QueryExpressionTable extends Table {
      */
     public final int getParameterOffset(ArrayList<Parameter> additionalParameters) {
         Query topQuery = getTopQuery();
-        int result = topQuery == null ? -1 : getMaxParameterIndex(topQuery.getParameters());
+        int result = topQuery == null ? 0 : Parameter.getMaxIndex(topQuery.getParameters());
         if (additionalParameters != null) {
-            result = Math.max(result, getMaxParameterIndex(additionalParameters));
+            result = Math.max(result, Parameter.getMaxIndex(additionalParameters));
         }
-        return result + 1;
+        return result;
     }
 
     @Override

--- a/h2/src/test/org/h2/test/db/TestRecursiveQueries.java
+++ b/h2/src/test/org/h2/test/db/TestRecursiveQueries.java
@@ -32,6 +32,7 @@ public class TestRecursiveQueries extends TestDb {
         testWrongLinkLargeResult();
         testSimpleUnionAll();
         testSimpleUnion();
+        testParameters();
     }
 
     private void testWrongLinkLargeResult() throws Exception {
@@ -176,6 +177,44 @@ public class TestRecursiveQueries extends TestDb {
         assertEquals(3, rs.getInt(1));
         assertFalse(rs.next());
 
+        conn.close();
+        deleteDb("recursiveQueries");
+    }
+
+    private void testParameters() throws Exception {
+        deleteDb("recursiveQueries");
+        Connection conn = getConnection("recursiveQueries");
+        PreparedStatement prep = conn.prepareStatement("WITH RECURSIVE T1(F1, F2) AS (\n" //
+                + "    SELECT CAST(? AS INT), CAST(? AS VARCHAR(15))\n" //
+                + "    UNION ALL\n" //
+                + "    SELECT (T1.F1 + CAST(? AS INT)), CAST((CAST(? AS VARCHAR) || T1.F2) AS VARCHAR(15))\n" //
+                + "    FROM T1 WHERE T1.F1 < 10\n" //
+                + "  ),\n" //
+                + "T2(G1, G2) AS (\n" //
+                + "    SELECT CAST(? AS INT), CAST(? AS VARCHAR(15))\n" //
+                + "    UNION ALL\n" //
+                + "    SELECT (T2.G1 + 1), CAST(('b' || T2.G2) AS VARCHAR(15))\n" //
+                + "    FROM T2 WHERE T2.G1 < 10\n" //
+                + "  )\n" //
+                + "SELECT T1.F1, T1.F2, T2.G1, T2.G2 FROM T1 JOIN T2 ON T1.F1 = T2.G1");
+        prep.setInt(1, 1);
+        prep.setString(2, "a");
+        prep.setInt(3, 1);
+        prep.setString(4, "a");
+        prep.setInt(5, 1);
+        prep.setString(6, "b");
+        ResultSet rs = prep.executeQuery();
+        StringBuilder a = new StringBuilder(10), b = new StringBuilder(10);
+        for (int i = 1; i <= 10; i++) {
+            a.append('a');
+            b.append('b');
+            assertTrue(rs.next());
+            assertEquals(i, rs.getInt(1));
+            assertEquals(a.toString(), rs.getString(2));
+            assertEquals(i, rs.getInt(3));
+            assertEquals(b.toString(), rs.getString(4));
+        }
+        assertFalse(rs.next());
         conn.close();
         deleteDb("recursiveQueries");
     }

